Fix tags being deleted by any batch action and by drap and drop reordering of records

ggppdk opened the pull request:
Pull Request for Issues #18317 and #18086 and #18057 

### Summary of Changes
The tags are not retrieved properly by the code of AdminModel class
(by the batch actions methods and saveorder() method)

This is to due wrongly calculated '**tableClassName**'
(when the **descendant** of this class is namespaced ?)
which results in empty UCM type & typeAlias, thus tags do not get loaded

There was common code (same code) called in 6 places 
- batch() method
- 4 batch*() methods
- saveorder() method

Fixed the code and moved it into a new method , initBatch()
(saveorder() needed some more fixing that the others)

### Testing Instructions
Use batch actions and manual reordering of records


### Expected result
The above task work properly 
and they also do not delete the tags assigned to the articles 


### Actual result
The tags are deleted


### Documentation Changes Required
initBatch() method added 

Should this be protected or private ?

ggppdk committed:
Fix tag not retrieved properly due to wrong className and wrong typeAlias

ggppdk commented:
This is nasty bug, maybe a 3.8.2 milestone ?

Quy commented on libraries/src/MVC/Model/AdminModel.php:
```diff
@@ -1478,4 +1435,34 @@ public function generateTitle($categoryId, $table)
+			$this->type = $this->contentType->getTypeByTable($this->tableClassName)
+				?: $this->contentType->getTypeByAlias($this->typeAlias);
+
+			// Get get tabs observer
```
Remove duplicate `get`

ghost commented:
I have tested this item :white_check_mark: successfully on caf8b822ecc84728ca73f6a648870ac3fcb8c275<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

ghost commented:
@ggppdk i guess this PR works only for com_content (not https://github.com/joomla-extensions/weblinks/issues/368)

wilsonge commented:
@laoneo

ggppdk commented:
> i guess this PR works only for com_content (not joomla-extensions/weblinks#368)

@franz-wohlkoenig 

This PR is a valid fix for all classes that extend ModelAdmin and face this issue,
I have tested just now with weblinks 3.7.0-beta1 and it is working

alikon commented on libraries/src/MVC/Model/AdminModel.php:
```diff
@@ -631,7 +590,7 @@ protected function batchTag($value, $pks, $contexts)
 				 * @var  \JTableObserverTags  $tagsObserver
 				 */
-				$tagsObserver = $table->getObserverOfClass('\JTableObserverTags');
+				$tagsObserver = $table->getObserverOfClass('Joomla\CMS\Table\Observer\Tags');;
```
remove double semicolon

alikon commented:
I have tested this item :white_check_mark: successfully on caf8b822ecc84728ca73f6a648870ac3fcb8c275<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

alikon commented:
but i'm afraid we still have some issues about classname and TagsObserver

old  `$tagsObserver = $table->getObserverOfClass('\JTableObserverTags');`
new `$tagsObserver = $table->getObserverOfClass('Joomla\CMS\Table\Observer\Tags');`

that is supposed to be fixed by #17739

ghost commented:
@alikon is this PR ready for RTC?

ggppdk committed:
Removed double colon

alikon commented:
i would  say yes cause it fix a big issue
but i suspect that there is still some issue with https://github.com/joomla/joomla-cms/pull/18328#issuecomment-336525833 under the hood

so i'm afraid we need some feedback from mantainers
perhaps setting RTC gain more priority

ggppdk commented:
@alikon 

yes it was supposed to have been fixed, as you said above,
but it seems that fix was incomplete

```php
foreach (JLoader::getDeprecatedAliases() as $alias)
```
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/observer/updater.php#L63

will return  '**jtableobservertags**' which is not equal to '**JTableObserverTags**'

thus this does not work
```php
$this->table->getObserverOfClass('\JTableObserverTags');
```

 (the slash is not the issue, lowercase comparison to uppercase is the issue)

alikon commented:
@ggppdk maybe this one #17834 unvalidate #17739

ggppdk commented:
The fixing of loading of classes via alias, regardless of letter case of the alias
will not effect this PR, and it should be done seperately of this PR

In this PR we just use the full qualified name of the class

This PR can be RTC

ghost commented:
RTC after two successful tests.

ggppdk committed:
Fixed duplicate word in comment

ggppdk commented:
@mbabker , @wilsonge

This is an important bug-fix of issues (of not uncommon usage) resulting in data loss (tags deleted)
please review for adding J3.8.2 milestone, and maybe reviewing the PR too

see brianteeman comment here also saying that this is urgent,
https://github.com/joomla/joomla-cms/issues/18317#issuecomment-336197306

aka if this, can be fixed properly then it should be in J3.8.2 ?

mbabker commented on libraries/src/MVC/Model/AdminModel.php:
```diff
@@ -1478,4 +1435,34 @@ public function generateTitle($categoryId, $table)
+		if (empty($this->batchSet))
+		{
+			$this->batchSet = true;
+			$this->user = \JFactory::getUser();
```
New class member variables should be documented for this and other `$this->foo` assignments.

ggppdk replied:
You are right, but the previous guy that originally added them, did not do it, 

you mean add them at the top of the class, together with other member variables,
and document them similarly ?

mbabker replied:
Yes please.

ggppdk replied:
Sure, I ll document them later today,
also 2 of them are not really needed to be defined as member variables

ggppdk committed:
Document 3 new class variables, and also update batchTag to use them

ggppdk committed:
Fix not defined local variable 'user' in last commit

ggppdk committed:
Fix not defined local variable 'batchSet'

ggppdk commented:
Added 3 new member variables and documented them,

but now this needs retesting !, no longer RTC !

ggppdk committed:
Better comments

ghost commented:
Status in Issue Tracker back on "Pending".

laoneo commented:
Are the three new global member variables used on other places as well or only for batch? If not then for me this can lead to confusion to have global member variables which are used only for a set of operations.

csthomas commented on libraries/src/MVC/Model/AdminModel.php:
```diff
@@ -1478,4 +1463,36 @@ public function generateTitle($categoryId, $table)
+
+		if ($batchSet === null)
+		{
+			$batchSet = true;
```
IMO It is not good to use static variables to store member variables.

Joomla probably does not use 2 instances of AdminModel in one request but unit tests may some day has a problem.

I do not know how many times `initBatch` is called per request, but If it is usual called  only 1 time per request then do not use any caching. Otherwise use `$this->batchSet` as it was before.

ggppdk replied:
You are right this should stay member variable,
for the case that 2 Models are created to execute batch actions
i ll commit fix

ggppdk commented:
@laoneo 

> Are  the three new global member variables used on other places as well or  only for batch? If not then for me this can lead to confusion to have  global member variables which are used only for a set of operations.

They are used for batch*() and for saveorder(), should i add this at the comments above their definitions ?

please note that
-- these are not new they already existed before this PR
-- their usage is now documented, before it was not

laoneo commented:
So if I call first batch* and then saveorder will it still work?

ggppdk committed:
Convert static variable to member variable, better comments

ggppdk commented:
> So if I call first batch* and then saveorder will it still work?

Yes all methods call `batchInit()` to populate the re-usable member variables

please note that the code regarding populating 
```php
$this->table
$this->type
$this->tagsObserver
```

existed before this PR,
just because same code was repeated 5 times, i moved it to seperate method

question 
these member variables were not documeted before

now i have renamed `$this->batchSet` to `$this->batchInitialized`
and removed `$this->user` and `$this->tableClassName`

is it a B/C break , should i restore them ?

csthomas commented:
May be you can use `$this->table === null` instead variable `$this->batchInitialized === null`?

csthomas commented:
> now i have renamed $this->batchSet to $this->batchInitialized
and removed $this->user and $this->tableClassName
>
> is it a B/C break , should i restore them ?

They have not been documented, I do not think so

ggppdk commented:
> May be you can use $this->table === null instead variable $this->batchInitialized === null?

yes my thought too,

just if some derived class decides to set $this->table in the constructor, then initBatch will not initialize the other member variables because $this->table === null will not be true

should i do this change ?

laoneo commented:
> is it a B/C break , should i restore them ?

If they have been not private then it can be counted as a BC break.

csthomas commented:
`$this->user` was used by `protected function checkCategoryId()`. As there can be more batch methods in subclasses then all old variables probably  has to stay initialised as before, some of them could be mark as deprecated. 

3rd party component may use one of that variable which is not documented above.
Does Joomla support B/C for such variables?

mbabker commented:
`$this->foo` when not declared on a class object is treated as a public variable.  So, there is a case to be made that removing the use of undeclared class variables counts as a B/C break because the variables were implicitly created through certain code paths.

ggppdk commented:
i ll add them back , exactly as before ))

ggppdk committed:
Restore variable names and make them member properties again to have B/C

ggppdk commented:
I have restored class member variables, now it should be B/C
Now it can be re-tested

csthomas commented on libraries/src/MVC/Model/AdminModel.php:
```diff
@@ -1478,4 +1481,37 @@ public function generateTitle($categoryId, $table)
+			$contentType = new \JUcmType;
+			$this->type = $contentType->getTypeByTable($this->tableClassName)
+				?: $contentType->getTypeByAlias($this->typeAlias);
+
```
To be strict I suggest to initialise `$this->contentType` as before. 

It would be nice to deprecate a few class variables (contentType, tableClassName, user) that you do not need right now if everyone agrees.

Patrickbmccabe commented:
I have this issue. (Please forgive my ignorance) How do I take part in a test ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

ghost commented:
@Patrickbmccabe please mark your Test as successfully:

- open [Issue Tracker](https://issues.joomla.org/tracker/joomla-cms/18328)
- Login with your github-Account
- Click on blue "Test this"-Button above Authors-Picture
- mark your Test as successfully
- hit "submit test result"

ggppdk committed:
Restore 1 more local variable as a member property

ggppdk commented:
Done i restored 1 more local variable as member property (as pointed out by @csthomas )

I retested too
Ready to be tested again

ggppdk commented:
@Puckmeister 
@franz-wohlkoenig 
@rgblogs
@Pilotnik 
@Fuchsritter 
@cybersalt
@AndySDH
@Shazrina1994

you had reported having issues with the tags being lost

maybe you can test this PR

ghost commented:
I have tested this item :white_check_mark: successfully on 88fec66f52603bed5c3fddde63b364858c85460c<br /><br />Darg-and-Drop-Reordering works also as Batch-Action; Test on Articles.

Thanks @ggppdk 
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

csthomas commented:
I have tested this item :white_check_mark: successfully on 88fec66f52603bed5c3fddde63b364858c85460c<br /><br />Tested #18317, #18086 and #18057<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

ghost commented:
RTC after two successful tests.

Pilotnik commented:
It's OK now. Thanks.

barraclm commented:
I would love to test this, but I am but a humble user, one of thousands, probably tens or even hundreds of thousands who uses Joomla to manage web sites. It may be 'fixed' for geeks, but for us lesser mortals it remains a MAJOR bug until an update appears in the update channel. Does anyone know when we might expect this?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

Patrickbmccabe commented:
I agree with Barraclm, I would love to give my time amd effort to this, but it is beyond my ability. This bug has caused major issue for a Library website I had created 100's of tags for. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18328">issues.joomla.org/tracker/joomla-cms/18328</a>.</sub>

ghost commented:
@barraclm https://github.com/joomla/joomla-cms/milestones - there you find expected Release Date.
@Patrickbmccabe https://docs.joomla.org/Bug_Tracking_Process - Information about Bug tracking using [Patchtester](https://github.com/joomla-extensions/patchtester).

barraclm commented:
Thank you Franz. That is very helpful.
